packages/core/core/src/Parcel.js

```diff
@@ -166,33 +181,35 @@ export default class Parcel {
   async run(): Promise<IBundleGraph<INamedBundle>> {
     let startTime = Date.now();
     if (!this.#initialized) {
-      await this.init();
+      await this._init();
     }
 
-    let result = await this.build({startTime});
-    await Promise.all([
-      this.#assetGraphBuilder.writeToCache(),
-      this.#runtimesAssetGraphBuilder.writeToCache(),
-    ]);
-
-    if (!this.#initialOptions.workerFarm) {
-      // If there wasn't a workerFarm passed in, we created it. End the farm.
-      await this.#farm.end();
-    }
+    let result = await this._build({startTime});
 
     if (result.type === 'buildFailure') {
       throw new BuildError(result.diagnostics);
     }
 
+    await this._end();
```

Comment:
So on buildFailure, we don't write anything to the cache?
Should `_end` be called before throwing the error?